### ISSUE TITLE
Nested coverage run fails

### ISSUE THREAD

**Issue by phst** [bazelbuild/rules_python#1394]
# 🐞 bug report

### Affected Rule

The issue is caused by the rule: `py_test`, `py_binary`

### Is this a regression?

Not that I'm aware

### Description

When running `bazel coverage` on a `py_test` that itself invokes a `py_binary`, the coverage run fails even though the test succeeds.

## 🔬 Minimal Reproduction

WORKSPACE:

```python
workspace(name = "test")

load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

http_archive(
    name = "rules_python",
    sha256 = "5868e73107a8e85d8f323806e60cad7283f34b32163ea6ff1020cf27abef6036",
    strip_prefix = "rules_python-0.25.0",
    url = "https://github.com/bazelbuild/rules_python/releases/download/0.25.0/rules_python-0.25.0.tar.gz",
)

load("@rules_python//python:repositories.bzl", "py_repositories", "python_register_toolchains")

py_repositories()

python_register_toolchains(
    name = "python_3_11",
    python_version = "3.11",
    register_coverage_tool = True,
)
```

BUILD:

```python
load("@python_3_11//:defs.bzl", "py_binary", "py_test")

py_binary(
    name = "bin",
    srcs = ["bin.py"],
)

py_test(
    name = "test",
    srcs = ["test.py"],
    data = [":bin"],
    deps = ["@rules_python//python/runfiles"],
)
```

bin.py:

```python
print('hello world')
```

test.py:

```python
import subprocess
from python.runfiles import runfiles
subprocess.run([runfiles.Create().Rlocation('test/bin')], check=True)
print('test binary finished')
```

Then run `bazel coverage :test`.

## 🔥 Exception or Error

Test output

```none
hello world
Wrote LCOV report to […]/testlogs/_coverage/test/test/pylcov.dat
test binary finished
Couldn't read '[…]/testlogs/_coverage/test/test/.coveragerc' as a config file
--
Coverage runner: Not collecting coverage for failed test.
The following commands failed with status 1
[…]/bin/test.runfiles/test/test
```

## 🌍  Your Environment

Debian GNU/Linux

**Output of `bazel version`:**

6.3.2

**Rules_python version:**

0.25.0

**Anything else relevant?**

I'm pretty sure what happens here is:

1. The wrapper for `test` writes (in the function `_RunForCoverage`) a temporary `.coveragerc` into `COVERAGE_DIR`.
2. The wrapper then invokes the `test.py` script.
3. `test.py` in turn invokes the wrapper for `bin.py`.
4. The wrapper for `bin` (in its own function `_RunForCoverage`) overwrites the `.coveragerc` written in step 1.
5. The wrapper for `bin` invokes `bin.py`.
6. `bin.py` runs and succeeds.
7. The wrapper for `bin` invokes `coverage lcov`.
8. The wrapper for `bin` deletes the `.coveragerc` file.
9. The wrapper for `bin` exits successfully.
10. `test.py` continues and finishes successfully.
11. The wrapper for `test` invokes `coverage lcov`, but that fails because the `.coveragerc` file is gone.
12. The wrapper for `test` exits with a failure status.

(So I guess this is more of a problem with Bazel itself since the wrapper script comes from there.)

Possible solutions:

- Don't create the `.coveragerc` file in `COVERAGE_DIR`, but in a private temporary directory.
- Recreate `.coveragerc` before invoking `coverage lcov`.

**Comment by rickeylev** [comment]
This must have been a pain to debug. Great analysis!

The only edge case I can think of with recreating the rcfile is if there are two concurrent programs being run. e.g. A calls B and C simultaneously. But, that would probably have other problems, too? I'm thinking the subsequent step to merge the lcov data might not work correctly. So I think its fine to not worry about that for now.

**Comment by aignas** [comment]
I think this is fixed by #1734 in bazel 7.0 and above. The CI is green and I am going to close the issue, please reopen if I have misunderstood something here.

**Comment by phst** [comment]
Looks like it's fixed at commit ebbcb6a548fbc1f62ec27a5aef6c815924ffbc66, thanks!
